tools/awslint/GUIDELINES.md

```diff
@@ -0,0 +1,425 @@
+
+> <a id="import-props-interface">awslint: import-props-interface</a>
+
+The "props" argument is `XxxImportProps`, which an interface that declares
+properties that allow the user to specify an external resource identity, usually
+by providing one or more resource attributes such as ARN, physical name, etc.
+
+The import interface should have the minimum required properties, that is: if it
+is possible to parse the resource name from the ARN (using `cdk.ArnUtils.parse`),
+then only the ARN should be required. In cases where it
+is not possible to parse the ARN (e.g. if it is a token and the resource name
+might have use "/" characters), both the ARN and the name should be optional and
+runtime-checks should be performed to require that at least one will be defined.
+See `ecr.RepositoryAttributes` for an example.
+
+The recommended way to implement the `import` method is as follows:
+
+1. A public abstract base class called `XxxBase` which implements `IXxx` and
+   extends `cdk.Construct`.
+2. The base class should provide as much of the implementation of `IXxx` as possible given the
+   context it has. In most cases, `grant` methods, `metric` methods, etc can be implemented at
+   at that level.
+5. A private class called `ImportedXxx` which extends `XxxBase` and implements
+   any remaining abstract members.
+4. The `import` static method should be have the following implementation:
+
+```ts
+public static import(scope: cdk.Construct, id: string, props: XxxImportProps): IXxx {
+  return new ImportedXxx(scope, id, props);
+}
+```
+
+## Exports
+
+> <a id="export">awslint: export</a>
+
+All resource interfaces (`IXxx`) must declare a method called `export` with the
+following signature:
+
+```ts
+export(): XxxImportProps
+```
+
+This method can be used to export this resource from the current stack and use
+it in another stack through an `Xxx.import` call.
+
+The implementation of `export` is different between internal resources (`Xxx`) and
+external imported resource (`ImportedXxx` as recommended above):
+
+For internal resources, the `export` method should produce a CloudFormation Output
+for each resource attribute, and return a set of `{ "Fn::ImportValue" }` tokens
+so they can be imported to another stack.
+
+```ts
+class Xxx extends XxxBase {
+  public export(): XxxImportProps {
+    return {
+      attr1: new cdk.Output(this, 'Attr1', { value: this.attr1 }).makeImportValue().toString(),
+      attr2: new cdk.Output(this, 'Attr2', { value: this.attr2 }).makeImportValue().toString(),
+    }
+  }
+}
+```
+
+For external resources, we know the actual values, so basically you would want to reflect
+your `props` as is:
+
+```ts
+class ImportedXxx extends XxxBase {
+  constructor(scope: cdk.Construct, id: string, private readonly props: XxxImportProps) {
+    // ...
+  }
+
+  public export() {
+    return this.props;
+  }
+}
+```
+
+The reason we are defining `export` on the resource interface and not on the resource
```

Comment:
I think it's worthwhile splitting this document into two parts:

1. These are the rules.
2. This is why the rules are what they are.

The reason is if I'm implementing something and I want to quickly validate whether I'm following the rules properly, I just want to be able to skim them and not have to skip swathes of rationale.

The rationale can probably go at the end or into a different file, it's going to be mostly interesting if people are curious or want to quibble.

Reply:
Do you think `awslint list` is not good enough?

```console
$ awslint list
module-name: module name must be @aws-cdk/aws-<namespace>
construct-ctor: signature of all construct constructors should be "scope, id, props"
resource-class: every resource must have a resource class (L2)
resource-class-is-construct: resource classes must extend "cdk.Construct" directly or indirectly
resource-props: an interface for resource props must be defined
resource-interface: every resource must have a resource interface
import-props-interface: every resource must have an "FooImportProps" interface
resource-interface-extends-construct: resource interface must extends cdk.IConstruct
resource-attribute: resources must represent all attributes as properties
resource-attribute-immutable: resource attributes must be immutable (readonly)
import: resource class must have a static "import" method
export: resource interface must have an "export" method
```

Reply:
I think there's tension between "rule caption", "rule + short description + example" and "rule + rationale". 

But also I don't want to create duplicate work for nothing. If we say this doc is the rationale doc, I think I can live with that.

Reply:
This is the rationale doc, but we can definitely add a longer description to the code itself.